diff --git a/dfs.cc b/dfs.cc
index b8b6b2d..e85e593 100644
--- a/dfs.cc
+++ b/dfs.cc
@@ -111,7 +111,9 @@ void* do_work(void* args)
         }//printf("\n %d",v);
 
         if(exist[v]==0)
-            continue;                              //if not in graph
+        {
+            continue;
+        }
          pthread_mutex_lock(&locks[v]);
          if(Q[v]==1)                       //if unset then set
            Q[v]=0;
@@ -119,10 +121,14 @@ void* do_work(void* args)
 
          for(int i = 0; i < edges[v]; i++)
          {
+            zsim_update_begin();
             int neighbor = W_index[v][i];
+            zsim_update_end();
 
             if(Q[neighbor]==0)
-              continue;
+            {
+                continue;
+            }
 
             stack[ptr] = neighbor;
             if(ptr<((largest+1)/P))
