diff --git a/bc.cc b/bc.cc
index f32f091..c01a478 100644
--- a/bc.cc
+++ b/bc.cc
@@ -131,7 +131,9 @@ void* do_work(void* args)
       {   //Bellman Ford type
          for(int i = 0; i < DEG; i++)
          {
+             zsim_update_begin();
              int Ewais1 = D[v]+W[v][i];
+             zsim_update_end();
              int Ewais2 = W_index[v][i];
              if((D[Ewais2] > Ewais1))
              {
